src/versionbits.cpp

```diff
@@ -1,55 +1,191 @@
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 16
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 17
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 18
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 19
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 20
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 21
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 22
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 23
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 24
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 25
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 26
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "",          // unallocated bit 27
+        /*.gbt_force =*/ false,
+    },
+    {
+        /*.name =*/ (char *) "testdummy", // unallocated bit 28
+        /*.gbt_force =*/ false,
     }
 };
+// bip135 end
+
 
+// bip135 begin
+// major adaptations to generalize and support new grace period parameters
 ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex* pindexPrev, const Consensus::Params& params, ThresholdConditionCache& cache) const
 {
     int nPeriod = Period(params);
     int nThreshold = Threshold(params);
     int64_t nTimeStart = BeginTime(params);
     int64_t nTimeTimeout = EndTime(params);
+    int nMinLockedBlocks = MinLockedBlocks(params);
+    int64_t nMinLockedTime = MinLockedTime(params);
+    int64_t nActualLockinTime = 0;
+    int nActualLockinBlock = 0;
+
+    if (nPeriod == 0)
+    {
+        // we cannot do anything further -this deployment is not really defined.
+        return THRESHOLD_DEFINED;
+    }
 
     // A block's state is always the same as that of the first of its period, so it is computed based on a pindexPrev whose height equals a multiple of nPeriod - 1.
     if (pindexPrev != NULL) {
+        assert(nPeriod);
         pindexPrev = pindexPrev->GetAncestor(pindexPrev->nHeight - ((pindexPrev->nHeight + 1) % nPeriod));
     }
 
-    // Walk backwards in steps of nPeriod to find a pindexPrev whose information is known
+    // Walk backwards in steps of nPeriod to find a pindexPrev which was DEFINED
     std::vector<const CBlockIndex*> vToCompute;
-    while (cache.count(pindexPrev) == 0) {
+    bool backAtDefined = (cache.count(pindexPrev) && cache[pindexPrev] == THRESHOLD_DEFINED);
```

Comment:
It had to do with the need to reliably determine nActualLockinBlock / nActualLockinTime during the ensuing forward walk.